### PR TITLE
add support for fan_speed on climate device

### LINKED ISSUE #466: Feature request: Add Fan mode and Swing mode to climate KNX devices
Hi All,
Would it be possible to add Fan Mode and Swing Mode to KNX Climate devices? It is supported by Home Assistant in other Climate integrations (e.g. Sensibo Climate). If needed I can help with testing, I have a KNX modules for Mitsubishi A/C units which enable to control all features of the A/C unit including Fan modes and Swing modes. If needed I can also share which DPTs are used to control those modes.

### PR DESCRIPTION
**Pull request by somdoron**
## Description

Add fan speed support to climate device. 

Fixes #466

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist

- [X] The documentation has been adjusted accordingly
- [X] Tests have been added that prove the fix is effective or that the feature works
- [X] The changes are documented in the changelog (docs/changelog.md)

### CODE REVIEW

File: docs/climate.md [hunk 1]
```diff
@@ -65,9 +68,12 @@ climate_mode = ClimateMode(
...
     group_address_controller_mode_state=None,
+    group_address_fan_speed=None,
+    group_address_fan_speed_state=None,
     operation_modes=None,
     controller_modes=None,
     device_updated_cb=None,
+    fan_max_step=None,
 )
 
 climate = Climate(
```
Comment (farmio) [inline docs/climate.md:79]: The new attributes should be added to `Climate`, not `ClimateMode`.

File: test/devices_tests/climate_test.py [hunk 2]
```diff
@@ -1562,3 +1562,78 @@ async def test_is_active(self):
...
+                payload=GroupValueWrite(DPTArray(140)),
+            )
+        )
+        assert climate_percent.current_fan_speed == 55
+
+        await climate_step.set_fan_speed(3)
+        assert xknx.telegrams.qsize() == 1
+        telegram = xknx.telegrams.get_nowait()
+        # 115 is 45% as byte (0...255)
+        assert telegram == Telegram(
```
Comment (farmio) [inline test/devices_tests/climate_test.py:1610]:
```suggestion
        await climate_step.set_fan_speed(3)
        assert xknx.telegrams.qsize() == 1
        telegram = xknx.telegrams.get_nowait()
        assert telegram == Telegram(
```

**Review by farmio — CHANGES_REQUESTED** [review]

**Comment by farmio** [comment]
Thank you very much! 👍

Do you plan to add this to the HA integration too?

I wonder if it would make sense to create a new device (from HA or in xknx.devices.Climate - like done for ClimateMode) to keep the climate module scope more clean -> 
Climate -> temperatures (and on/off 😬)
ClimateMode -> modes and presets
ClimateFan -> fan speed, fan modes, swing modes and whatever else a climate fan does... (maybe even xknx.devices.Fan can be reused for this) https://developers.home-assistant.io/docs/core/entity/climate
What do you think?

File: xknx/devices/climate.py [hunk 3]
```diff
@@ -135,6 +140,33 @@ def __init__(
             after_update_cb=self.after_update,
         )
 
+        self.fan_speed: RemoteValueDptValue1Ucount | RemoteValueScaling
+        self.fan_mode = FanSpeedMode.STEP if fan_max_step else FanSpeedMode.PERCENT
+        self.fan_max_step = fan_max_step
```
[the remark this answers is not in the document]
Comment (somdoron) [reply xknx/devices/climate.py:145]: I removed the field. I can add those checks, but I want to keep it similar to how fan device works.

**Comment by somdoron** [comment]
I do plan to add this to Home Assistant. I don't have a strong opinion on how to integrate it, but from Home Assistant's perspective, fan speed is part of the climate device, so it makes sense to handle it the same way here. Additionally, the "switch" for the fan speed is actually the climate switch, so I wouldn't use the fan device for controlling the climate fan speed.

As for using a different class for ClimateFan, I think it would be relatively thin compared to ClimateMode, but I'm fine with moving it to its own class.

**Comment by farmio** [comment]
> I do plan to add this to Home Assistant.

Nice 👍 let me know if any questions arise.

>  Additionally, the "switch" for the fan speed is actually the climate switch

Yes sure, but afaik the "switch" of the xknx Fan device is optional. I don't have a strong opinion about using this vs. a new class vs. putting it in xknx Climate (and I personally don't own any fans 🙃 whatsoever)

> I think it would be relatively thin compared to ClimateMode

... which is probably a good thing 😛 Was just looking at #466 and thought there could be more to come regarding vane stuff. But could of course always be refactored later on if needed.

Looking at HA climate entity documentation, it seems they don't support percentage fan speeds, but only modes like `[LOW, MEDIUM, HIGH]` (or custom). Do you already have plans on how to integrate that?

**Review by farmio — APPROVED** [review]

**Review by marvin-w — APPROVED** [review]
Nice! Thanks!

**Comment by adorobis** [comment]
> Fixes #466

Great work. Does implement also the swing mode (as described in the feature request) or only the fan speed?

**Comment by farmio** [comment]
@adorobis this PR only adds fan speed.

@somdoron I usually implement my HA integration changes and tests locally with a non-released version of xknx so I can still make additions or changes if necessary. Let me know if you want to do the same - and if so, when it is ready for release so we can publish (which will be needed in order to file the HA core PR).

**Pull request merged**